internal/testhelpers/session.go

```diff
@@ -136,92 +178,261 @@ func NewHTTPClientWithSessionToken(t *testing.T, reg *driver.RegistryDefault, se
+// `SessionLifespanAndPrivilegedMaxAgeProvider` for setting the session lifespan and privileged age.
+//
+// Example Usage:
+//
+//	func TestSessionBehavior(t *testing.T) {
+//		var unprivilegedClient *http.Client
+//		unprivilegedClient = NewHTTPClientBuilder(t).
+//			SetReqestFromWhoAmI().
+//			SetIdentityFromNew().
+//			SetSessionWithProvider(UnrivilegedProvider()).
+//			ClientWithSessionToken()
+//	}
+func (c *HTTPClientBuilder) SetSessionWithProvider(provider *SessionLifespanAndPrivilegedMaxAgeProvider) *HTTPClientBuilder {
+	c.nilRequestCheck()
+	c.nilIdentityCheck()
+	session, err := session.NewActiveSession(
+		c.request,
+		c.identity,
+		provider,
 		time.Now(),
 		identity.CredentialsTypePassword,
 		identity.AuthenticatorAssuranceLevel1,
 	)
-	require.NoError(t, err, "Could not initialize session from identity.")
-
-	return NewHTTPClientWithSessionToken(t, reg, s)
+	require.NoError(c.t, err, "Could not initialize session from identity.")
+	return c.SetSession(session)
 }
 
-func EnsureAAL(t *testing.T, c *http.Client, ts *httptest.Server, aal string, methods ...string) {
-	res, err := c.Get(ts.URL + session.RouteWhoami)
-	require.NoError(t, err)
-	sess := x.MustReadAll(res.Body)
-	require.NoError(t, res.Body.Close())
-	assert.EqualValues(t, aal, gjson.GetBytes(sess, "authenticator_assurance_level").String())
-	for _, method := range methods {
-		assert.EqualValues(t, method, gjson.GetBytes(sess, "authentication_methods.#(method=="+method+").method").String())
+// Record a callback function for mutating the session after the client has been created. This
+// is more useful than recording the session pointer because mutations to the session wouldn't
+// be persistend unless the session was then passed to the private `testhelpers.maybePersistSession`
+// function. This function compromises in giving access to the session without uncontrollably
+// exposing the persistence function.
+//
+// Example Usage:
+//
+//	func TestSessionBehavior(t *testing.T) {
+//		// Define a var at which we'll record the session mutation function
+//		var sessionMutator func(mutateFunc func(session *session.Session))
+//
+//		// Construct a client with a default unprivileged session
+//		client := NewHTTPClientBuilder(t).
+//			SetReqestFromWhoAmI().
+//			SetIdentityFromNew().
+//			SetSessionDefault(UnrivilegedProvider()).
+//			RecordSessionMutator(reg, &sessionMutator).
+//			ClientWithSessionToken()
+//
+//		// Run some tests using `client` requiring the session to be unprivileged
+//		...
+//
+//		// Pretend the session has been refreshed, and is now privileged for an additional minute
+//		sessionMutator(func(session *session.Session) {
+//			session.SetPrivilegedUntil(time.Now().Add(time.Minute))
+//		})
+//
+//		// Run some tests using `client` requiring the session to be privileged
+//		...
+//	}
+func (c *HTTPClientBuilder) RecordSessionMutator(reg *driver.RegistryDefault, funcRef *func(func(session *session.Session))) *HTTPClientBuilder {
```

Comment:
I figured this function could be confusing to come across in the few instances where we us it, so I wanted to make sure I provided some detailed docs on its usage and purpose for existing.